src/main/java/org/testng/GuiceHelper.java

```diff
@@ -121,12 +129,34 @@ public static Injector createInjector(ITestContext context,
     Class<? extends IModuleFactory> factory = guice.moduleFactory();
     if (factory != IModuleFactory.class) {
       IModuleFactory factoryInstance = parentInjector.getInstance(factory);
-      Module moduleClass = factoryInstance.createModule(context, testClass);
-      if (moduleClass != null) {
-        result.add(moduleClass);
+      Module module = factoryInstance.createModule(context, testClass);
+      if (module != null) {
+        result.add(module);
       }
     }
-
+    result.addAll(getSpiModules());
     return result;
   }
-}
+
+  private List<Module> getSpiModules() {
```

Comment:
@bartoszpop - I dont think we need to do all the double check idiom. We should be able to resort to doing lock free lazy loading by just working with inner classes as explained in [this](https://stackoverflow.com/a/8297830) StackOver flow post. You just need to define an inner class that works as a singleton and which internally takes care of loading the modules. Something like below

```java
  private List<Module> getSpiModules() {
    return SpiModuleLoader.getInstance().getModules();
  }

  private static class SpiModuleLoader {
    private static final SpiModuleLoader instance = new SpiModuleLoader();

    public static SpiModuleLoader getInstance() {
      return instance;
    }

    private final List<Module> modules = new LinkedList<>();
    private SpiModuleLoader() {
      for (IModule module : ServiceLoader.load(IModule.class)) {
        modules.add(module);
      }
    }

    public List<Module> getModules() {
      return modules;
    }
  }
```

Reply:
Done